### PR TITLE
fix: statement.cancel() waits for statement to finish

### PR DESCRIPTION
**Pull request by crystall-bitquill**
### Summary

Fix statement.cancel() waiting for statement to finish

### Description

When a long running query is executed and `statement.cancel()` is used to stop it, the driver will not immediately cancel the statement and will wait for it to finish instead. This PR replaces the synchronized block with a ReentrantLock.
Addresses #527

### CODE REVIEW

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/ConnectionProxy.java [hunk 1]
```diff
@@ -347,14 +357,23 @@ public Object invoke(Object proxy, Method method, Object[] args) throws Throwabl
...
-      synchronized(this.invokeOn) {
+      if (NON_SYNCHRONIZED_METHODS.contains(methodName)) {
+        this.LOCK.lock();
+      }
+      try {
         Object result =
             ConnectionProxy.this.pluginManager.execute(
                 this.invokeOnClass,
-                methodName,
+                method.getName(),
```
Comment (sergiyvamz) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/ConnectionProxy.java:367]: Why methodName isn't good in this case?

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/ConnectionProxy.java [hunk 2]
```diff
@@ -347,14 +357,23 @@ public Object invoke(Object proxy, Method method, Object[] args) throws Throwabl
...
                 () -> method.invoke(this.invokeOn, args),
                 args);
         return proxyIfReturnTypeIsJdbcInterface(method.getReturnType(), result);
+      } finally {
+        try {
+          this.LOCK.unlock();
```
Comment (sergiyvamz) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/ConnectionProxy.java:373]: Do we need to check is lock has been acquired above?

**Review by sergiyvamz — APPROVED** [review]

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/ConnectionProxy.java [hunk 3]
```diff
@@ -313,6 +315,14 @@ private boolean isDirectExecute(String methodName) {
   class JdbcInterfaceProxy implements InvocationHandler {
     private final Object invokeOn;
     private final Class<?> invokeOnClass;
+    private final ReentrantLock LOCK = new ReentrantLock();
```
Comment (karenc-bq) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/ConnectionProxy.java:318]: Variable name shouldn't be all uppercase here.

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/ConnectionProxy.java [hunk 4]
```diff
@@ -313,6 +315,14 @@ private boolean isDirectExecute(String methodName) {
...
+
+    private final Set<String> NON_SYNCHRONIZED_METHODS =
```
Comment (karenc-bq) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/ConnectionProxy.java:320]: Same as above

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/ConnectionProxy.java [hunk 5]
```diff
@@ -347,14 +357,23 @@ public Object invoke(Object proxy, Method method, Object[] args) throws Throwabl
         return executeMethodDirectly(methodName, args);
       }
 
-      synchronized(this.invokeOn) {
+      boolean locked = false;
+      if (nonSynchronizedMethods.contains(methodName)) {
+        this.lock.lock();
+        locked = true;
```
Comment (karenc-bq) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/ConnectionProxy.java:363]: What is the purpose of this flag

Comment (karenc-bq) [reply src/main/user-impl/java/com/mysql/cj/jdbc/ha/ConnectionProxy.java:363]: You can use the built in checks, either `isHeldByCurrentThread` or `isLocked`

**Review by sergiyvamz — APPROVED** [review]

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/ConnectionProxy.java [hunk 6]
```diff
@@ -313,6 +315,14 @@ private boolean isDirectExecute(String methodName) {
   class JdbcInterfaceProxy implements InvocationHandler {
     private final Object invokeOn;
     private final Class<?> invokeOnClass;
+    private final ReentrantLock lock = new ReentrantLock();
```
[the remark this answers is not in the document]
Comment (sergiyvamz) [reply src/main/user-impl/java/com/mysql/cj/jdbc/ha/ConnectionProxy.java]: The lock can't be static!

**Pull request merged**